### ISSUE TITLE
Make simplebrowser example run

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtwebengine-os2#6]
This is a more specific continuation of #4.

The release build (run with QTWEBENGINE_CHROMIUM_FLAGS=--single-process) just shows an empty web page so far. The debug build fails with this:

```
[46441:4:0814/011119.786000:ERROR:database.cc(1808)] Cookie sqlite error 14, errno -2: unable to open database file, sql: -- sqlite3_open()
[46441:4:0814/011119.859000:ERROR:sqlite_persistent_cookie_store.cc(773)] Unable to open cookie DB.
[46441:4:0814/011119.868000:ERROR:database.cc(1808)] Cookie sqlite error 14, errno -2: unable to open database file, sql: -- sqlite3_open()
[46441:4:0814/011119.868000:ERROR:sqlite_persistent_cookie_store.cc(773)] Unable to open cookie DB.
[46441:1:0814/011119.999000:ERROR:time_zone_monitor_os2.cc(30)] Not implemented reached in static std::unique_ptr<device::TimeZoneMonitor> device::TimeZoneMonitor::Create(scoped_refptr<base::SequencedTaskRunner>)
failed to open /dev/urandom: No such file or directory
[46441:5:0814/011120.573000:ERROR:debugger_posix.cc(176)] Not implemented reached in bool base::debug::BeingDebugged()
Received signal 6
[end of stack trace]
Calling _exit(1). Core file will not be generated.
```

**Comment by dmik** [comment]
The /dev/urandom problem lies in BoringSSL. It needs more porting to work properly on OS/2. Will fix it in a separate ticket.

**Comment by dmik** [comment]
With `/dev/urandom` fixed, I get further up to this:
![изображение](https://user-images.githubusercontent.com/900726/90316188-296ec580-df29-11ea-8211-49972fef5946.png)
Then it starts loading a page but eventually crashes with the following fatal error:
```
[14830:16:0815/195733.246000:FATAL:memory.cc(22)] Out of memory. size=16384
```
I need to look at that closer.

Other suspicious things I see are these:
```
[14830:6:0815/195636.465000:ERROR:database.cc(1808)] Cookie sqlite error 14, errno -2: unable to open database file, sql: -- sqlite3_open()
[14830:6:0815/195636.471000:ERROR:database.cc(1808)] Cookie sqlite error 14, errno -2: unable to open database file, sql: -- sqlite3_open()
...
[14830:6:0815/195636.986000:ERROR:cert_verify_proc_builtin.cc(463)] No net_fetcher for performing AIA chasing.
[14830:5:0815/195636.990000:ERROR:ssl_client_socket_impl.cc(962)] handshake failed; returned -1, SSL error code 1, net_error -202
```
I'm not sure yet if they are critical or not.

**Comment by dmik** [comment]
With the debug build (see https://github.com/bitwiseworks/libc/issues/82#issuecomment-677618883) I quickly found out where it fails with OOM. It's `bitmap_allocation::AllocateMappedBitmap` in `components/vis/common/resources/bitmap_allocation.cc`. What happens there is PlatformMemoryHandle gets passed over the Mojo (IPC) framework via serialize/deserialze (it seems to be used even in single process mode) and that somehow doesn't work on OS/2. Should be easy to fix. All other similar cases like that will start working after that I suppose.

**Comment by dmik** [comment]
After fixing some more stuff within https://github.com/bitwiseworks/qtwebengine-chromium-os2/issues/14, I got further:

![Снимок экрана 2020-08-23 в 18 11 26](https://user-images.githubusercontent.com/900726/90982118-b939f280-e56d-11ea-90b1-d62b9c9575b3.png)

As you can see, it can already render page contents  — which looks promising! Meaning that rendering generally works. It even picks up WarpSans already. But there are still some more problems with resource allocation (ERR_INSUFFICIENT_RESOURCES). Needs further debugging/fixing.

**Comment by dmik** [comment]
(Note that the defect on the screen shot is due to some screen shot capture problem, it crashed right after showing this due to some failed assertion check.

**Comment by dmik** [comment]
SQLite problems are there because Chromium uses its own version (`third_party/sqlite`) and there is no option to use the system library because they claim to have patches that are never going to be accepted upstream. So the only solution here is to apply OS/2 patches there. Will create a ticket for that.

**Comment by dmik** [comment]
Fixing SQLite brought the Cookies database but ERR_INSUFFICIENT_RESOURCES is still there. Need to deep further.

Also noticed this in the output:
```
[65229:16:0824/230143.556000:ERROR:backend_impl.cc(1063)] Critical error found -8
[65229:16:0824/230143.566000:WARNING:storage_block-inl.h(143)] Failed data load.
```
I don't know if it's related or not — it doesn't happen upon each run (but each run ends with ERR_INSUFFICIENT_RESOURCES).

**Comment by dmik** [comment]
From what I see in the detailed debug logs, networking works per se. It sends a valid request:
```
[53121:5:0825/180819.206000:VERBOSE1:spdy_session_key.cc(29)] SpdySessionKey(host=www.ru:80, proxy=direct://, privacy=0
[53121:5:0825/180819.493000:VERBOSE1:http_stream_parser.cc(235)] SendRequest() request_line = "GET / HTTP/1.1

" headers = "Host: www.ru

Connection: keep-alive

Upgrade-Insecure-Requests: 1

User-Agent: Mozilla/5.0 (Unknown; OS/2 i386) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.13.1 Chrome/73.0.3683.105 Safari/537.36

Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8

Accept-Encoding: gzip, deflate



"
```
and gets a valid response from the server:
```
[53121:5:0825/180819.759000:VERBOSE1:http_stream_parser.cc(1040)] ParseResponseHeaders() content_length = "-1
" headers = "HTTP/1.1 200 OK
Server: nginx/1.18.0
Date: Tue, 25 Aug 2020 14:08:20 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Powered-By: Express
ETag: W/"38e8-vV6gjpzl0z1VYh+zOuUI7gk+GQ0"
Content-Encoding: gzip
X-Kong-Upstream-Latency: 178
X-Kong-Proxy-Latency: 1
Via: kong/1.4.1
X-RID: a36725c2489b73c246aaefdb7f0001b5
"
```
But then some part of the code spits that ERR_INSUFFICIENT_RESOURCES and aborts page loading. Still looking for it.

**Comment by dmik** [comment]
So, I did it. This last problem was `SharedMemory::GetHandleLimit` that would return a too big number on OS/2 and it would get turned into zero maximum allowed in-flight requests per one renderer process. As a result, `ResourceDispatcherHostImpl::HasSufficientResourcesForRequest` would return false and `MojoAsyncResourceHandler::OnWillRead` in turn would abort the page load with ERR_INSUFFICIENT_RESOURCES. Fixing that made loading pages basically work. Some proofs:

![Снимок экрана 2020-08-25 в 22 50 22](https://user-images.githubusercontent.com/900726/91226832-3b711500-e72e-11ea-8535-d51257c1bb89.png)
![Снимок экрана 2020-08-25 в 22 52 02](https://user-images.githubusercontent.com/900726/91226837-3d3ad880-e72e-11ea-82f0-50fda81770eb.png)
![Снимок экрана 2020-08-25 в 22 55 14](https://user-images.githubusercontent.com/900726/91226840-3e6c0580-e72e-11ea-8f69-0a58956e1449.png)

**Comment by dmik** [comment]
Note that it's not very stable yet — a bunch of assertions makes it crash in the debug build at least. Some of them:
```
[57201:12:0826/013109.391000:FATAL:page_allocator.cc(204)] Check failed: !(reinterpret_cast<uintptr_t>(address) & kPageAllocationGranularityOffsetMask). 
...
[57159:12:0826/012542.065000:FATAL:options_validation.h(35)] Check failed: options && IsAligned<MOJO_ALIGNOF(Options)>(options). 
...
[55450:16:0825/233745.365000:FATAL:histogram.cc(1096)] Check failed: false. 
...
[46076:12:0823/224311.877000:FATAL:l10n_util.cc(734)] Check failed: std::string::npos != pos (4294967295 vs. 4294967295) Didn't find a $1 placeholder in Check your proxy settings or contact your network administrator to
      make sure the proxy server is working. If you don't believe you should
      be using a proxy server:
      $1
...
[55641:12:0825/235854.995000:ERROR:SkFontMgr_FontConfigInterface.cpp(221)] ../../../../../qt5/qtwebengine/src/3rdparty/chromium/third_party/skia/src/ports/SkFontMgr_FontConfigInterface.cpp:221: fatal error: "Not implemented."
```
None of them look serious. Will fix it in the next days.

**Comment by dmik** [comment]
With the above problems fixed/worked around, the app crashes because of some problems in JIT assembly generated for some JavaScript when visiting some JS-extensive sites (like youtube.com). Will create a ticket for that too.

**Comment by dmik** [comment]
Looking at html5test results, we're not that bad actually, better than the current Firefox (not the old version 45 we have on OS/2, but the current upstream one). And just slightly worse than Chrome, perhaps due to some missing OS integration:
![Снимок экрана 2020-08-31 в 13 34 40](https://user-images.githubusercontent.com/900726/91711824-3cbb9b00-eb8f-11ea-99f0-64bb808fa58f.png)
![Снимок экрана 2020-08-31 в 13 35 55](https://user-images.githubusercontent.com/900726/91711828-3f1df500-eb8f-11ea-80e4-7f90abd71d4e.png)

**Comment by dmik** [comment]
With the JS crash sorted out in https://github.com/bitwiseworks/qtwebengine-chromium-os2/issues/21, JS crashes are gone and the next problem is OOM condition which I'm frequently getting now in various scenarios (YouTube included). I need to debug this OOM more to see what's causing it.

**Comment by dmik** [comment]
BTW, besides OOM from Chromium itself, I'm also getting ENOMEM from `pthread_create` which originates from `_beginthread`. The reason for that is not yet known to me, need to debug `_beginthread` in LIBC (and that's always not easy).

**Comment by dmik** [comment]
I tracked down the first OOM. It happens in `ClientDiscardableSharedMemoryManager::AllocateLockedDiscardableSharedMemory` which eventually tries to map a `PlatformSharedMemoryRegion` instance via its `MapAtInternal` call. This call takes a shared memory address (a memory "handle" in terms of Chromium) and attempts to create an alias for it with `DosAliasMem`. Why it is done so, is described in detail in https://github.com/bitwiseworks/qtwebengine-chromium-os2/issues/14. Seems that `DosAliasMem` can only create aliases in low memory and that one gets exhausted pretty fast. When the crash happens, it wants to map 4MB and there is only 800K in private low, so no surprise `DosAliasMem` fails with ERROR_NOT_ENOUGH_MEMORY. (Private high still contains about 2.2G of free RAM).

Looks like we have reached the boundaries of our limited memory allocation implementation need a better solution as described in https://github.com/bitwiseworks/qtwebengine-chromium-os2/issues/14 — by introducing a new memory allocation API that will allow us to not leak memory and to get rid of DosAliasMem completely so that original memory objects will be reused w/o wasting address space. This will take a few days.
